executor/builder.go

```diff
@@ -1476,8 +1489,35 @@ func (b *executorBuilder) buildTableDual(v *plannercore.PhysicalTableDual) Execu
 	return e
 }
 
-// `getSnapshotTS` returns the timestamp of the snapshot that a reader should read.
+// `getSnapshotTS` returns for-update-ts if in insert/update/delete/lock statement otherwise snapshot-read-ts
 func (b *executorBuilder) getSnapshotTS() (uint64, error) {
+	if b.forDataReaderBuilder {
+		return b.dataReaderTS, nil
+	}
+
+	if b.inInsertStmt || b.inUpdateStmt || b.inDeleteStmt || b.inSelectLockStmt {
+		return b.getForUpdateTS()
+	}
+
+	return b.getSnapshotReadTS()
+}
+
+// getSnapshotReadTS returns the ts used by insert/update/delete and select for update
+func (b *executorBuilder) getForUpdateTS() (uint64, error) {
+	ts, err := b.getSnapshotReadTS()
```

Comment:
The semantics of these terms of the TS names still looks confusing to me. Will it be better after your later refactoring?

Reply:
Yes. I'll refactor it in the later refactoring. But I think I can change the method `getSnapshotReadTS` to `getIsolateTS` and returns the no-update ts according to isolation level.